File: app/src/main/java/com/stedi/randomimagegenerator/app/model/data/GeneratorType.java
package com.stedi.randomimagegenerator.app.model.data;

import android.support.annotation.NonNull;
import android.support.annotation.StringRes;

import com.stedi.randomimagegenerator.app.R;

import java.util.ArrayList;
import java.util.List;

public enum GeneratorType {
    COLORED_CIRCLES(false, R.string.generator_colored_circles),
    COLORED_RECTANGLE(false, R.string.generator_colored_rectangle),
    COLORED_PIXELS(false, R.string.generator_colored_pixels),
    FLAT_COLOR(false, R.string.generator_flat_color),
    COLORED_NOISE(false, R.string.generator_colored_noise),
    MIRRORED(true, R.string.effect_mirrored),
    TEXT_OVERLAY(true, R.string.effect_text_overlay);

    private final boolean isEffect;
    private final int stringRes;

    private static GeneratorType[] effectTypes;
    private static GeneratorType[] nonEffectTypes;

    GeneratorType(boolean isEffect, int stringRes) {
        this.isEffect = isEffect;
        this.stringRes = stringRes;
    }

    public boolean isEffect() {
        return isEffect;
    }

    @StringRes
    public int getStringRes() {
        return stringRes;
    }

    @NonNull
    public static GeneratorType[] nonEffectTypes() {
        if (nonEffectTypes == null) {
            List<GeneratorType> list = new ArrayList<>();
            for (GeneratorType type : GeneratorType.values()) {
                if (!type.isEffect())
                    list.add(type);
            }
            nonEffectTypes = list.toArray(new GeneratorType[list.size()]);
        }
        return nonEffectTypes;
    }

    @NonNull
    public static GeneratorType[] effectTypes() {
        if (effectTypes == null) {
            List<GeneratorType> list = new ArrayList<>();
            for (GeneratorType type : GeneratorType.values()) {
                if (type.isEffect())
                    list.add(type);
            }
            effectTypes = list.toArray(new GeneratorType[list.size()]);
        }
        return effectTypes;
    }
}

File: app/src/main/java/com/stedi/randomimagegenerator/app/model/data/GeneratorType.kt
package com.stedi.randomimagegenerator.app.model.data

enum class GeneratorType(val isEffect: Boolean, val isEditable: Boolean) {
    RANDOM_NON_EFFECT(false, false),
    FLAT_COLOR(false, false),
    COLORED_RECTANGLE(false, true),
    COLORED_CIRCLES(false, true),
    COLORED_LINES(false, true),
    COLORED_PIXELS(false, true),
    COLORED_NOISE(false, true),
    MIRRORED(true, false),
    THRESHOLD(true, false),
    TEXT_OVERLAY(true, false);

    companion object {
        val EFFECT_TYPES: Array<GeneratorType> by lazy {
            GeneratorType.values().filter { it.isEffect }.toTypedArray()
        }

        val NON_EFFECT_TYPES: Array<GeneratorType> by lazy {
            GeneratorType.values().filter { !it.isEffect }.toTypedArray()
        }
    }
}

fun GeneratorType?.effectOrdinal() = GeneratorType.EFFECT_TYPES.indexOf(this)

fun GeneratorType?.nonEffectOrdinal() = GeneratorType.NON_EFFECT_TYPES.indexOf(this)